src/main/scala/mesosphere/marathon/experimental/storage/BucketingPersistenceStore.scala

```diff
@@ -0,0 +1,178 @@
+  * packet containing all the children names. Some experimentation showed that each child element in that packet has ~4bytes
+  * overhead for encoding. So, let's say each child znode *name* e.g. holding Marathon app definition is 50 characters long
+  * (seems like a good guess given 3-4 levels of nesting e.g. `eng_dev_databases_my-favourite-mysql-instance`), we could only
+  * have: 4096 * 1024 / (50 + 4) =  ~75k children nodes until we hit the exception.
+  *
+  * This trait solves the above mentioned limitations by putting the elements into buckets. Each element is then
+  * stored in `/$basePath/$bucket/$relative/element/path`. This way the total number of possible elements is
+  * NumberOfBuckets * MaxElementsPerBucket where later parameter depends on the length of children znode name (see
+  * calculation above)
+  * Every element passed should implement [[BucketNode]] trait that defines the hash of the bucket where the element is stored
+  * and the relative path below.
+  *
+  * Same goes for reading elements - one need to pass a [[BucketPath]] which defines bucket hash along with the element
+  * relative path.
+  *
+  * Note: this class isn't used anywhere yet and serves more as a demonstration.
+  *
+  * @param base base path to all stored elements e.g. "/templates"
+  * @param store instance of the underlying [[ZooKeeperPersistenceStore]]
+  * @param MaxBuckets number of buckets to use. It hould never be changed in production since all data location will
+  *                   change and re-hashing will be needed
+  */
+class BucketingPersistenceStore(store: ZooKeeperPersistenceStore, base: String, MaxBuckets: Int = 16) {
+
+  /**
+    * Map of bucket numbers -> bucket name e.g. (5 -> "05")
+    */
+  def buckets: Map[Int, String] = (0 to MaxBuckets).map(b => b -> f"$b%02d")(collection.breakOut)
+
+  /**
+    * Return bucket name calculated from the passed bucket hash.
+    */
+  def bucket(bucketNode: BucketNode): String = bucket(bucketNode.bucketPath)
+  def bucket(bucketPath: BucketPath): String = buckets(bucketPath.bucketHash % buckets.size)
+
+  /**
+    * Return full path for the passed [[BucketNode]]/[[BucketPath]] combining base path, bucket name and relative path.
+    */
+  def path(bucketNode: BucketNode): String = path(bucketNode.bucketPath)
+  def path(bucketPath: BucketPath): String = Paths.get(base, bucket(bucketPath), bucketPath.relativePath).toString
+
+  /**
+    * Helper stages, mapping from BucketNode/BucketPath to Node/Path
+    */
+  def toNode: Flow[BucketNode, Node, NotUsed] = Flow[BucketNode].map(bn => Node(path(bn), bn.payload))
+  def toPath: Flow[BucketPath, String, NotUsed] = Flow[BucketPath].map(bp => path(bp))
+
+  // All methods below wrap the storage methods by mapping from BucketNode/BucketPath to Node/Path
+  def create(bucketNode: BucketNode): Future[String] = store.create(Node(path(bucketNode), bucketNode.payload))
```

Comment:
What do you think about some documentation for the return type?

Reply:
Removed